--- uboot-socfpga/include/configs/socfpga_common.h	2015-01-16 15:51:03.650585434 -0800
+++ uboot-socfpga-new/include/configs/socfpga_common.h	2015-01-18 19:27:28.383726931 -0800
@@ -223,6 +223,7 @@
 	"fpga=0\0" \
 	"fpgadata=0x2000000\0" \
 	"fpgadatasize=0x700000\0" \
+	"soc_workshop_use_ltsi=0\0" \
 	CONFIG_KSZ9021_CLK_SKEW_ENV "=" \
 		__stringify(CONFIG_KSZ9021_CLK_SKEW_VAL) "\0" \
 	CONFIG_KSZ9021_DATA_SKEW_ENV "=" \
